0024 and 0025: message validation and transform (#81)
emqx/eip: active/0024--message-validation-and-transform.md

> # Message validation and transform

Author: Renamed the title to "Built-in Message Validation".

The decision so far is to keep the two features separate. Validation is more independent.
EIP-0025 drafts the Message Transform feature which will be kept closer to the rule-engine.

## Changelog

* 2020-10-21: @zmstone Initial draft

## Abstract

A new feature for EMQX v5 to validate, transform, and filter MQTT messages.

## Motivation

### Data validation

EMQX has no native support for data validation.

Technically, by using EMQX rules engine, it is possible to validate the incoming
MQTT messages before sending the message off to data bridges.

For instance, one can write an rules engine SQL like below to
drop messages which are not avro encoded and also when the 'val' field is not greater than 10.

```
SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
```

However, currently rules are only used in data bridges, but it does not interfere
MQTT message pub/sub.
For example, if there is subscriber subscribing to `t/#` it would still receive the avro encoded message.

If we also want to stop the message being published, then there would be a need for a message republish action,
and the publisher and subscriber will likely be forced to use different topics.
e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.

> This can work, but not easy to use.

Reviewer: Random thought, you probably have already pondered it. We could introduce something like _rewrite_ action in the Rule Engine and grant it an ability to erase the original message and replace it with one or more other messages, or even terminate the connection if configured accordingly.

Author: Can maybe put this whole new thing to `rule_engine.validation_and_transform`.
However mixing it with existing actions would mean
1. Actions are managed using different APIs, it's probably not the best way to manage validation/transformation
2. Rules lack of ordering, if we are to force the such actions to be positioned in front of the list then it's effectively a separate list.

Reviewer: Very good points. A couple of questions / thoughts.

> Actions are managed using different APIs

Q: Different from which?

> Rules lack of ordering, if we are to force the such actions to be positioned in front of the list

I can very well stray away from the initial requirements, but assuming this feature is for some advanced user, then it's easy to imagine that they want to control the order as well, i.e.: 1) log and/or archive in some remote system the message completely as it is, 2) validate / transform it, 3) push through some bridge. However this starts to sound more like _pipelines_ honestly, I'm not sure if we are ready to go that far.

Anyway, I'm just exploring / expressing design ideas, please feel free to ignore them.

Author: Actions have a `actions` config root, and they are so far dedicated to data integration, and managed using the `/actions` HTTP endpoint. It's would be odd if we create a `/validations` but put things under the `actions` space.

Actions can have order when pipelined and used in rule, but rules do not have order and it would be hard to define.
To support data validation as a rule, it would imply to create a singleton rule and chained (ordered actions)

Reviewer: > To support data validation as a rule, it would imply to create a singleton rule and chained (ordered actions)

Do you imagine here a scenario where a customer wants that "ALL incoming messages should be validated / filtered / transformed first", before even hitting the broker (e.g. routing / hooks)? Then I agree, it sounds impossible to achieve on the data integration side alone. As a sidenote: that sounds a lot like ingress bridge + rules but for the MQTT clients themselves (i.e. not for some remote system).

However, if hitting the broker is not an issue then I don't quite understand why a singleton rule is a must. Again, a sufficiently advanced customer may have a lot of different clients with different patterns, which may require different validation / transformation applied. Thus, different pipelines.

But I agree that ordering between rules is totally unnecessary design complexity.

### Message transform

Same as data validation, if message is to be transformed using rules engine SQL,
it would require a republish action.

## Design

Develop a data processing feature which can support:

- Message validation
- Message transform
- Message filtering

This data processing step should happen after authorization and before rules engine.
Probably can make use of the 'message.publish' with a hook priority which
positions the callback after authorization, but before data bridges.

### High level design

* Introduce 2 different data processing types: `validation` and `tranform`.

> * Make use of the rules engine SQL, but without the `FROM` caluse as it's always the MQTT message being th input.

Reviewer 2: ```suggestion
* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being the input.
```

The each data process rule should have a unique ID. And each rule should have a input topic-filter name which
can be used to build a topic index for quick look-up.

The common parts can be described as hocon config below:

```
{
    id = id1 # unique id assigned by user
    topic = 't/1' # input
    sql = "SELECT ....." # processing
    type = validation | transform
}

```

### Message Validation

```
{
    id = id1
    topic = 't/1'
    type = validation

>     action_if_invalid = disconnect # or 'drop'

Reviewer 3: If we use `message filtering` as common denominator for both validation and filtering, we could have instead `action_if_passed` and `action_if_failed`.

```
{
    id = id2
    topic = 't/2'
    type = tranform
    sql = "SELECT payload.value + 1 as new.payload.value"
}
```

To make it easer for user, we should also consider:

- Support all the existing message properties binding, for example `topic`, `qos` `retain`.
- Use `new.user_properties` for user properties.
  This is because the user properties in MQTT message object quite deeply nested,
  so it was not quite easy to use.
  We are already using `user_properties` keyword in some data bridge actions.

To support message expansion (one transform to many),
the `new` will be copied as many times as a `FOREACH` statement has to loop.

```
sql = "FOREACH payload.myarray as element DO element as new.payload
```

## Configuration Changes

New config which may look like:

```
message_process = [
    {
        description = "drop message if it is not compabitble with my avro schema"
        topic = 't/1'
        type = validation
        action_if_invalid = disconnect
        sql = "SELECT schema_decode(my_avro_schema, payload) as decoded WHERE decoded != undefined"
    },
    {
        description = "increment payload.value"
        topic = 't/2'

>         type = tranform

Reviewer 3: ```suggestion
        type = transform
```

>         sql = "SELECT * as new, payload.value + 1 as new.payload.value"

Reviewer: My 2 cents. Given that Rule SQL semantics in general provide a DSL to describe how to transform 1 message into 0 or 1 results (or even ≥0 with `FOREACH`), it's not obvious why this type distinction is needed. If running some SQL over a message produces 0 results (or fails to compute) it filters it out in some sense, otherwise transforms. Besides, with this type distinction the SQL syntax in different contexts might need to be constrained (e.g. disallow `FOREACH` in filters, disallow `WHERE` in transforms), which can be unintuitive for the user.

Author: do you mean there is no need for 'transform' vs. 'validation' ?
  --- it's mostly for config clarity, e.g. when it's `validation`, there there must be `action_if_invalid`.

or do you mean there is no need for 'transform' vs. 'filter'? 
  --- I don't think 'filter' type is proposed in this doc. "message filter" is just one of the input requirements, it's not mapped to a process type.

Reviewer: > do you mean there is no need for 'transform' vs. 'validation' ?

Mostly this. Agree that config clarity is important. I'm just trying to throw some random ideas that could help to _reduce_ the number of new concepts.

Another random thought though: if we want to _disconnect_ the client when it sends garbage, maybe it's the job for the authorization layer?

Author: > Another random thought though: if we want to disconnect the client when it sends garbage, maybe it's the job for the authorization layer?

Before this EIP, there were some discussions around:
1. Extend authz checks to support SQL  --- I like this idea very much ;)
2. Extend rule engine to support authz-like actions --- Product team proposed this, and I think @terry-xiaoyu had also came to this according to his comment.

This EIP is kind of striking in the middle because it's an important feature and deserves it's own space.

    }
]
```

## Backwards Compatibility

New feature.

## Document Changes

> Docs should be created to guid users to use the new feature.

Reviewer 3: ```suggestion
Docs should be created to guide users to use the new feature.
```
